clippy_lints/src/unused_async.rs

```diff
@@ -38,7 +40,23 @@ declare_clippy_lint! {
     "finds async functions with no await statements"
 }
 
-declare_lint_pass!(UnusedAsync => [UNUSED_ASYNC]);
+#[derive(Default)]
+pub struct UnusedAsync {
+    /// Keeps track of async functions used as values (i.e. path expressions to async functions that
+    /// are not immediately called)
+    async_fns_as_value: LocalDefIdSet,
+    /// Functions with unused `async`, linted post-crate after we've found all uses of local async
+    /// functions
+    unused_async_fns: FxHashMap<LocalDefId, UnusedAsyncFn>,
```

Comment:
There's no reason for this to be a `HashMap` instead of a `Vec`. Using a `Vec` will also give a stable lint order.

Reply:
> There's no reason for this to be a `HashMap` instead of a `Vec`. Using a `Vec` will also give a stable lint order.

Oh right, I used a `HashMap` because I was originally thinking that I could remove defs from `unused_async_fns` as we find paths, but realized that it's probably not really worth it.
But what do you mean about a stable lint order? Do you mean that we wouldn't need to do it post-crate?

Reply:
The order of items iterated on for a hashmap is essentially random, I believe it's stable for `FxHashMap` between different runs but if a function's added the order will completely change (which makes the diff pretty annoying)

Reply:
Stable in this case meaning stable between different builds of clippy and adding an item doesn't rearrange other items are linted in. Changes to the internals of the `HashMap` should not cause clippy's tests to fail.

In the future if you do need to use lint from a `HashMap` you can sort by the `HirId` emitting the lint.

Reply:
That makes sense, thanks for clarifying